### PR TITLE
FASM for ice40

### LINKED ISSUE #468: Integrate ice40 FASM output
splitting #288

### PR DESCRIPTION
**Pull request by elmsfu**
- Add fasm metadata
 - Add tools for converting between FASM and ASC

Fixes #468

### CODE REVIEW

**Comment by elmsfu** [comment]
TODO:
 - [ ] Expand tests
 - [ ] remove HLC
 - [ ] get xml validation passing

HX8k is crashing vpr and I need bisect that first to make sure I'm not missing anything on that path when I fixed a few issue with 1k parts.

File: ice40/primitives/sb_ff/sb_ff.pb_type.xml [hunk 1]
```diff
@@ -53,6 +53,8 @@
         # no_setreset
         # --
     </meta>
+    <meta name="fasm_features">
```
Comment (mithro) [inline ice40/primitives/sb_ff/sb_ff.pb_type.xml]: Does this need something in here?

File: ice40/primitives/sb_ff/sb_ff.pb_type.xml [hunk 2]
```diff
@@ -146,6 +155,11 @@
         # use_cen
         async_setreset
         set_noreset
+    </meta>
+       <meta name="fasm_features">
```
Comment (mithro) [inline ice40/primitives/sb_ff/sb_ff.pb_type.xml]: Indenting

File: ice40/cells/plb/plb.pb_type.xml [hunk 3]
```diff
@@ -164,14 +164,26 @@
  </pb_type>
 
  <pb_type name="BLK_IG-LUTFF" num_pb="8">
-  <xi:include href="../lutff/lutff.pb_type.xml" xpointer="xpointer(pb_type/child::node())"/>
+   <xi:include href="../lutff/lutff.pb_type.xml" xpointer="xpointer(pb_type/child::node())"/>
+   <metadata>
+     <meta name="hlc_cell">lutff</meta>
```
Comment (litghost) [inline ice40/cells/plb/plb.pb_type.xml]: Did you intend to leave the hlc_cell annotation?

File: ice40/utils/fasm_icebox/asc2fasm.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python3
+import sys
+from fasm_icebox_utils import ascToFasm
```
Comment (litghost) [inline ice40/utils/fasm_icebox/asc2fasm.py]: PEP8 violation?  ascToFasm should be asc_to_fasm?

Comment (litghost) [reply ice40/utils/fasm_icebox/asc2fasm.py]: https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
+"""Convert the icestorm database to a prjxray type DB
```
Comment (mithro) [inline ice40/utils/fasm_icebox/fasm_icebox_utils.py]: This file needs to be formatted I think?

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+
+def _check_iceconfig_entry(tile, entry):
+  """
+  check an entry is set for a tile
+  """
+  bm, bp = _get_iceconfig_bits(tile, entry[0])
+  return np.all(bp[bm] == tile[bm])
+
+
+class featureAccumulator(dict):
```
Comment (litghost) [inline ice40/utils/fasm_icebox/fasm_icebox_utils.py]: PEP8 naming violation?

File: ice40/utils/fasm_icebox/ice40_feature.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    def __init__(self, tile_type, loc, bit_tuples, parts, idx=None):
+        self.tile_type = tile_type
+        self.loc = loc
+        self.bit_tuples = bit_tuples
+        self.parts = parts
+        self.idx = idx
+
+    @staticmethod
+    def _parse_icedb_bit(bit):
```
Comment (litghost) [inline ice40/utils/fasm_icebox/ice40_feature.py]: Why static method versus a free function?

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+ - break out bits for LC_[0-7] INIT
+
+
+Another name variant
+
+HLC introduce a globally unique name for each wire named on the right endpoint.
+FASM will have a slight variation for some features due to not supporting / character
+
+"""
+import numpy as np
```
Comment (litghost) [inline ice40/utils/fasm_icebox/fasm_icebox_utils.py]: numpy doesn't seem needed here?  Can you explain why it is being used over plain python types?

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py (hunk 6 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Done

File: ice40/utils/fasm_icebox/asc2fasm.py (hunk 4 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/asc2fasm.py]: Done

File: ice40/cells/plb/plb.pb_type.xml (hunk 3 above)
Comment (elmsfu) [reply ice40/cells/plb/plb.pb_type.xml]: I will remove in follow up PR.

File: ice40/utils/fasm_icebox/ice40_feature.py (hunk 7 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/ice40_feature.py]: I was scoping it to as a helper to the class, but moved it out if someone wants to use it elsewhere.

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py (hunk 8 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: I think it simplifies the code using numpy arrays. I'd actually like to push this down to icestorm. icebox stores the bits as a list of strings which this goes back and forth between. If you feel strongly I switch to 2D lists.

File: ice40/primitives/sb_ff/sb_ff.pb_type.xml (hunk 1 above)
Comment (elmsfu) [reply ice40/primitives/sb_ff/sb_ff.pb_type.xml]: removed empty features

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py (hunk 5 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Done.

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+    def test_asc_to_fasm(self):
+        with open('tt','w') as f:
+            res = asc_to_fasm('test_data/test1.asc', f)
+        with open('t2', 'w') as f:
+            fasm_to_asc('tt', f)
+    """
+
+
+if __name__ == "__main__":
+    import argparse
```
Comment (litghost) [inline ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Move this code to a main function

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 10]
```diff
@@ -0,0 +1,596 @@
...
+        if tile_type == "RAMB" and feature.parts[-1].startswith("INIT"):
+            _array_to_tile(tile_bits, tile, is_hex=True)
+        else:
+            _array_to_tile(tile_bits, tile)
+
+    # TODO: would be nice to upstream a way to write to non-files
+    ic.write_file(outf.name)
+
+
+class TestConversion(unittest.TestCase):
```
Comment (litghost) [inline ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Should the test case be in the file itself?

File: ice40/utils/fasm_icebox/ice40_feature.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        fe = self.toFasmEntry()
+        rem = ".".join(self.parts)
+        return (
+            "{0:s}_X{1[0]:d}_Y{1[1]:d}.{2:s}{3:s}".format(
+                self.tile_type, self.loc, rem, self.bit_str
+            ),
+            fasm_bits,
+        )
+
+    def toIceDbStr(self):
```
Comment (litghost) [inline ice40/utils/fasm_icebox/ice40_feature.py]: Consider raising an exception or deleting this method?

File: ice40/primitives/sb_ff/sb_ff.pb_type.xml (hunk 2 above)
Comment (elmsfu) [reply ice40/primitives/sb_ff/sb_ff.pb_type.xml]: Done

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 12]
```diff
@@ -0,0 +1,596 @@
+"""Convert the icestorm database to a prjxray type DB
+
+The goal is that the icestorm db doesn't encode information about BRAM init or
```
[the remark this answers is not in the document]
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Done

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 13]
```diff
@@ -0,0 +1,596 @@
...
+import re
+import unittest
+import numpy as np
+
+import icebox
+import fasm
+from ice40_feature import Feature, IceDbEntry, FasmEntry
+
+
+def _nibbles_to_bits(line):
```
[the remark this answers is not in the document]
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Done

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Done

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 14]
```diff
@@ -0,0 +1,596 @@
...
+    Class to help accumulate iCE40 features for output as FASM file or a FASM DB
+    """
+
+    def append_feature(self, feature):
+        self[feature.toFasmEntry().feature] = feature
+
+    def append_ice_entry(self, tile_type, tile_loc, bits, names, idx, negate=False):
+
+        if negate:
+            feature = Feature.fromIceDbEntry(
```
[the remark this answers is not in the document]
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Done

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py (hunk 10 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Not sure if we have a convention for how we want to do this, but I did move it to a separate file similar to rr_graph

File: ice40/utils/fasm_icebox/ice40_feature.py (hunk 11 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/ice40_feature.py]: Done

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py (hunk 9 above)
Comment (elmsfu) [reply ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Done

File: ice40/primitives/sb_ff/sb_ff.pb_type.xml (hunk 2 above)
Comment (elmsfu) [reply ice40/primitives/sb_ff/sb_ff.pb_type.xml]: Done

File: ice40/utils/fasm_icebox/fasm_icebox_utils.py [hunk 15]
```diff
@@ -0,0 +1,541 @@
...
+    if is_hex:
+        array = np.array([_nibbles_to_bits(line) for line in tile], dtype=bool)
+    else:
+        array = np.array([[int(xx) for xx in line] for line in tile], dtype=bool)
+    return array
+
+
+def _array_to_tile(tile_bits, tile, is_hex=False):
+    """
+    convert a numpy array to text icedb tile
```
Comment (litghost) [inline ice40/utils/fasm_icebox/fasm_icebox_utils.py]: Docstring convention is to have the summary on the first line of the string:
https://www.python.org/dev/peps/pep-0257/

You have yours on the second line.

**Comment by litghost** [comment]
Wait for green Travis, and I think we are ready to merge?

File: ice40/CMakeLists.txt [hunk 16]
```diff
@@ -17,6 +17,25 @@ function(icestorm_setup)
...
   get_target_property(LIBFTDI_TARGET env LIBFTDI_TARGET)
 
+  add_conda_pip(
+    NAME numpy
+    NO_EXE
+    )
+
+  add_thirdparty_package(
+    NAME fasm
+    BUILD_INSTALL_COMMAND "cd ${symbiflow-arch-defs_SOURCE_DIR}/third_party/fasm && ${PYTHON3} setup.py install"
```
Comment (mithro) [inline ice40/CMakeLists.txt]: `python setup.py develop` might be a better option here.

See this -> https://stackoverflow.com/questions/19048732/python-setup-py-develop-vs-install
> python setup.py install is used to install (typically third party) packages that you're not going to develop/modify/debug yourself.

**Pull request merged**